### PR TITLE
testsuite: cannot case-switch on MPI handles 2

### PR DESCRIPTION
**Pull request by jeffhammond**
## Pull Request Description

Change another case-switch to if-else, because the MPI standard does not state that MPI_Datatype handles are compile-time constants.

This is the same as https://github.com/pmodels/mpich/pull/6011.

## Author Checklist
* [x] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [x] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [ ] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [x] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

### CODE REVIEW

File: test/mpi/coll/allred.c [hunk 1]
```diff
@@ -267,26 +267,26 @@ static void set_index_power(MPI_Datatype mpi_type, int category, void *arr, int
...
+        SET_PAIR_CASE(int, MPI_2INT)
+        else
+        SET_PAIR_CASE(long, MPI_LONG_INT)
+        else
+        SET_PAIR_CASE(short, MPI_SHORT_INT)
+        else
+        SET_PAIR_CASE(float, MPI_FLOAT_INT)
+        else
+        SET_PAIR_CASE(double, MPI_DOUBLE_INT)
+        else { assert(0); }
```
Comment (hzhou) [inline test/mpi/coll/allred.c:289]: This is ugly :) I would rather spell the macros out. There are only 5 cases anyway. Curiously, is there a system missing one of these type? `int`, `long`, `short`, `float`, `double` are standard C types, right?

Comment (hzhou) [reply test/mpi/coll/allred.c:289]: Alternatively, define the macro for the case body, then
```
if (mpi_type == MPI_2INT) {
    SET_PAIR(int);
} else if (mpi_type == MPI_LONG_INT) {
    SET_PAIR(long);
} ...
```
Rationale: the `if-else if - else` is a semantic unit, it is bad to split in and out of macros.

Comment (hzhou) [reply test/mpi/coll/allred.c:289]: Also, the spell checker doesn't work with macros that can't follow `;`.

Comment (jeffhammond) [reply test/mpi/coll/allred.c:289]: Sure, I was merely trying to pattern match the previous version as much as possible.  I'll go with the `SET_PAIR` macro idea above.

File: test/mpi/coll/allred.c [hunk 2]
```diff
@@ -267,26 +267,28 @@ static void set_index_power(MPI_Datatype mpi_type, int category, void *arr, int
...
 
-#define SET_PAIR_CASE(type, mpi_type_) \
-    if ( mpi_type == mpi_type_ ) \
+#define SET_PAIR(type) \
         { \
             struct type ## _test *pp = p; \
             pp->a = val_a; \
             pp->b = val_b; \
-        } \
+        }
```
Comment (hzhou) [inline test/mpi/coll/allred.c:275]: Suggest use the form -
```
    do { \
        ...
    } while (0)
```

Some compilers dislike `{ ... };`

**Comment by hzhou** [comment]
For your reference, here is what I have in my `.bashrc`:
```
function mpich_whitespace {
    REPO=$PWD
    pushd $HOME/work/pull_requests
    if test -d mpich-whitespace ; then
        rm -rf mpich-whitespace
    fi

    echo git clone mpich-github mpich-whitespace ...
    git clone $REPO mpich-whitespace
    cd mpich-whitespace
    git config core.whitespace blank-at-eol,blank-at-eof,space-before-tab,tab-in-indent

    git fetch origin
    git checkout $1
    git rebase --verbose --exec maint/hooks/pre-commit origin/main
    popd
}

function mpich_spellcheck {
    REPO=$PWD
    pushd $HOME/work/pull_requests
    if test -d mpich-spellcheck ; then
        rm -rf mpich-spellcheck
    fi

    echo git clone mpich-github mpich-spellcheck ...
    git clone $REPO mpich-spellcheck
    cd mpich-spellcheck

    git fetch origin
    git checkout $1
    git rebase --verbose --exec maint/hooks/spellchecker origin/main
    popd
}

```
You'll need adjust the paths and repo names.

**Comment by jeffhammond** [comment]
I wrote this on MacOS.  I wonder if that codes works with ZSH.

**Comment by hzhou** [comment]
> I wrote this on MacOS. I wonder if that codes works with ZSH.

You'll find out :)

**Comment by raffenet** [comment]
@jeffhammond should we update your git author email?
```
Author: Jeff Hammond <jeff.r.hammond@intel.com>
Date:   Wed Nov 16 11:21:29 2022 +0200
```

**Comment by jeffhammond** [comment]
Yes please. I was confused about that and unsure where it came from.

**Comment by raffenet** [comment]
```
$ git config --global user.name "John Doe"
$ git config --global user.email johndoe@example.com
```
Or edit your `~/.gitconfig` directly. I can take care of this commit. What should it be now?

**Comment by raffenet** [comment]
test:mpich/ch4/ofi

**Comment by raffenet** [comment]
test:mpich/ch4/ofi

**Review by hzhou — APPROVED** [review]
LGTM

**Pull request merged**